File: Frontend/src/app/services/book.ts
/*
 * Bookapp - Book Management System Frontend
 * Copyright (c) 2025 Jorge Avila
 * Author: Jorge Avila (jorgeavilas@icloud.com)
 * Repository: https://github.com/Jojje84/Bookapp
 * License: MIT License - see LICENSE file for details
 * 
 * This file contains the book service with CRUD operations.
 * Original work by Jorge Avila - please maintain attribution.
 */

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../app.config'; 
import { Observable } from 'rxjs';

export interface Book {
  id?: number;
  title: string;
  author: string;
  publishedDate: string;
}

@Injectable({
  providedIn: 'root'
})
export class BookService {
  private baseUrl = environment.apiUrl;  // 👈 Render-backend

  constructor(private http: HttpClient) {}

  // 📚 Hämta alla böcker
  getBooks(): Observable<Book[]> {
    return this.http.get<Book[]>(`${this.baseUrl}/api/books`);
  }

  // 📖 Hämta en bok
  getBook(id: number): Observable<Book> {
    return this.http.get<Book>(`${this.baseUrl}/api/books/${id}`);
  }

  // ➕ Lägg till en bok
  addBook(book: Book): Observable<Book> {
    return this.http.post<Book>(`${this.baseUrl}/api/books`, book);
  }

  // ✏️ Uppdatera en bok
  updateBook(id: number, book: Book): Observable<void> {
    return this.http.put<void>(`${this.baseUrl}/api/books/${id}`, book);
  }

  // 🗑️ Ta bort en bok
  deleteBook(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/api/books/${id}`);
  }
}

File: Backend/Models/Book.cs
/*
 * Bookapp - Book Management System Backend
 * Copyright (c) 2025 Jorge Avila
 * Author: Jorge Avila (jorgeavilas@icloud.com)
 * Repository: https://github.com/Jojje84/Bookapp
 * License: MIT License - see LICENSE file for details
 * 
 * This file contains the Book model with validation attributes.
 * Original work by Jorge Avila - please maintain attribution.
 */

using System.ComponentModel.DataAnnotations;

namespace BookApi.Models
{
    public class Book
    {
        public int Id { get; set; }

        [Required(ErrorMessage = "Titel är obligatorisk")]
        [MinLength(2, ErrorMessage = "Titel måste vara minst 2 tecken")]
        public string Title { get; set; }

        [Required(ErrorMessage = "Författare är obligatorisk")]
        [MinLength(2, ErrorMessage = "Författare måste vara minst 2 tecken")]
        public string Author { get; set; }

        [Required(ErrorMessage = "Publiceringsdatum är obligatoriskt")]
        [DataType(DataType.Date, ErrorMessage = "Publiceringsdatum måste vara ett giltigt datum")]
        public DateTime PublishedDate { get; set; }
    }
}
